src/main/java/org/gridsuite/gateway/TokenValidatorGlobalPreFilter.java

```diff
@@ -0,0 +1,108 @@
+ */
+@Component
+public class TokenValidatorGlobalPreFilter implements GlobalFilter {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(TokenValidatorGlobalPreFilter.class);
+
+    @Value("${allowed-issuers}")
+    private List<String> allowedIssuers;
+
+    @Autowired
+    private final GatewayService gatewayService;
+
+    public TokenValidatorGlobalPreFilter(GatewayService gatewayService) {
+        this.gatewayService = gatewayService;
+    }
+
+    @Override
+    public Mono<Void> filter(ServerWebExchange exchange, GatewayFilterChain chain) {
+        LOGGER.debug("checking authorization");
+        List<String> ls = exchange.getRequest().getHeaders().get("Authorization");
+        if (ls == null) {
+            LOGGER.debug("Authorization header is required");
+            // set UNAUTHORIZED 401 response and stop the processing
+            exchange.getResponse().setStatusCode(HttpStatus.UNAUTHORIZED);
+            return exchange.getResponse().setComplete();
+        }
+
+        LOGGER.debug("checking issuer");
+        String authorization = ls.get(0);
+        String token = authorization.split(" ")[1];
+        JWT jwt;
+        JWTClaimsSet jwtClaimsSet;
+        try {
+            jwt = JWTParser.parse(token);
+            jwtClaimsSet = jwt.getJWTClaimsSet();
+        } catch (java.text.ParseException e) {
+            // Invalid plain JOSE object encoding
+            LOGGER.debug("Invalid plain JOSE object encoding");
+            // set UNAUTHORIZED 401 response and stop the processing
+            exchange.getResponse().setStatusCode(HttpStatus.UNAUTHORIZED);
+            return exchange.getResponse().setComplete();
+        }
+
+        if (allowedIssuers.stream().noneMatch(iss -> jwtClaimsSet.getIssuer().startsWith(iss))) {
+            LOGGER.debug(jwtClaimsSet.getIssuer() + " Issuer is not in the issuers white list");
+            // set UNAUTHORIZED 401 response and stop the processing
+            exchange.getResponse().setStatusCode(HttpStatus.FORBIDDEN);
+            return exchange.getResponse().setComplete();
+        }
+
+        try {
+            JWT idToken = JWTParser.parse(token);
+            Issuer iss = new Issuer(jwt.getJWTClaimsSet().getIssuer());
+            ClientID clientID = new ClientID(jwt.getJWTClaimsSet().getAudience().get(0));
+
+            JWSAlgorithm jwsAlg = JWSAlgorithm.parse(jwt.getHeader().getAlgorithm().getName());
+            URL jwkSetURL = null;
+
+            jwkSetURL = new URL(gatewayService.getJwksUrl(jwt.getJWTClaimsSet().getIssuer()));
+
+            // Create validator for signed ID tokens
+            IDTokenValidator validator = new IDTokenValidator(iss, clientID, jwsAlg, jwkSetURL);
+
+            IDTokenClaimsSet claims = validator.validate(idToken, null);
+            // we can safely trust the JWT
+            LOGGER.debug("Token verified, it can be trusted");
+            return chain.filter(exchange);
+        } catch (Exception e) {
```

Comment:
catch more specific exceptions

Reply:
For example if the http requests to get the keys fail, we should return a 500.
if the chain.filter throws an exception we shouldn't catch it